{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "88d3745d",
   "metadata": {},
   "source": [
    "The goal of this notebook is to take a group of three patients that were clustered very close to each other after hierarchical clustering and figure out what they have in common. This will help us figure out what is causing their specific cases of cancer."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "d56c1b89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import libraries\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b379e729",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in information for normal patients\n",
    "\n",
    "%store -r normal_patients\n",
    "%store -r normal_patient_mutations\n",
    "%store -r normal_prot\n",
    "%store -r candidate_cnv_patients"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a28ab06f",
   "metadata": {},
   "source": [
    "These three patient IDs come from the dendrogram after PCA and hierarchical clutsering. They were very close to each other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "b7597f23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>HGNC_UniProt_ID(supplied_by_UniProt)</th>\n",
       "      <th>HGNC_Ensembl_ID(supplied_by_Ensembl)</th>\n",
       "      <th>HGNC_UCSC_ID(supplied_by_UCSC)</th>\n",
       "      <th>Oreganno_Build</th>\n",
       "      <th>Simple_Uniprot_alt_uniprot_accessions</th>\n",
       "      <th>dbSNP_TOPMED</th>\n",
       "      <th>HGNC_Entrez_Gene_ID(supplied_by_NCBI)</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>getz</th>\n",
       "      <th>washu</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>BCHE</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>p.F385del</td>\n",
       "      <td>590.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr3</td>\n",
       "      <td>165829880</td>\n",
       "      <td>165829882</td>\n",
       "      <td>+</td>\n",
       "      <td>DEL</td>\n",
       "      <td>...</td>\n",
       "      <td>P06276</td>\n",
       "      <td>ENSG00000114200</td>\n",
       "      <td>uc003fem.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A8K7P8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>590.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>FLG</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.E2287K</td>\n",
       "      <td>2312.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr1</td>\n",
       "      <td>152308027</td>\n",
       "      <td>152308027</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>P20930</td>\n",
       "      <td>ENSG00000143631</td>\n",
       "      <td>uc001ezu.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q01720|Q5T583|Q9UC71</td>\n",
       "      <td>0.99991239806320081,0.00001592762487257,0.0000...</td>\n",
       "      <td>2312.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>MFSD9</td>\n",
       "      <td>Intron</td>\n",
       "      <td>NaN</td>\n",
       "      <td>84804.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr2</td>\n",
       "      <td>102732136</td>\n",
       "      <td>102732136</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>Q8NBP5</td>\n",
       "      <td>ENSG00000135953</td>\n",
       "      <td>uc002tcb.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q4ZG89|Q53TU0|Q96GQ4|Q9BRI8</td>\n",
       "      <td>0.99995221712538226,0.00004778287461773</td>\n",
       "      <td>84804.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>TRPC1</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R707C</td>\n",
       "      <td>7220.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr3</td>\n",
       "      <td>142806074</td>\n",
       "      <td>142806074</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>P48995</td>\n",
       "      <td>ENSG00000144935</td>\n",
       "      <td>uc003evb.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q14CE4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7220.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>UGDH</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>p.R142H</td>\n",
       "      <td>7358.0</td>\n",
       "      <td>hg38</td>\n",
       "      <td>chr4</td>\n",
       "      <td>39510701</td>\n",
       "      <td>39510701</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>...</td>\n",
       "      <td>O60701</td>\n",
       "      <td>ENSG00000109814</td>\n",
       "      <td>uc003guk.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B3KUU2|B4DN25|O60589</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7358.0</td>\n",
       "      <td>BRCA</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 131 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name         Gene           Mutation   Location  Entrez_Gene_Id NCBI_Build  \\\n",
       "Patient_ID                                                                   \n",
       "01BR023      BCHE       In_Frame_Del  p.F385del           590.0       hg38   \n",
       "01BR023       FLG  Missense_Mutation   p.E2287K          2312.0       hg38   \n",
       "01BR023     MFSD9             Intron        NaN         84804.0       hg38   \n",
       "01BR023     TRPC1  Missense_Mutation    p.R707C          7220.0       hg38   \n",
       "01BR023      UGDH  Missense_Mutation    p.R142H          7358.0       hg38   \n",
       "\n",
       "Name       Chromosome  Start_Position  End_Position Strand Variant_Type  ...  \\\n",
       "Patient_ID                                                               ...   \n",
       "01BR023          chr3       165829880     165829882      +          DEL  ...   \n",
       "01BR023          chr1       152308027     152308027      +          SNP  ...   \n",
       "01BR023          chr2       102732136     102732136      +          SNP  ...   \n",
       "01BR023          chr3       142806074     142806074      +          SNP  ...   \n",
       "01BR023          chr4        39510701      39510701      +          SNP  ...   \n",
       "\n",
       "Name       HGNC_UniProt_ID(supplied_by_UniProt)  \\\n",
       "Patient_ID                                        \n",
       "01BR023                                  P06276   \n",
       "01BR023                                  P20930   \n",
       "01BR023                                  Q8NBP5   \n",
       "01BR023                                  P48995   \n",
       "01BR023                                  O60701   \n",
       "\n",
       "Name       HGNC_Ensembl_ID(supplied_by_Ensembl)  \\\n",
       "Patient_ID                                        \n",
       "01BR023                         ENSG00000114200   \n",
       "01BR023                         ENSG00000143631   \n",
       "01BR023                         ENSG00000135953   \n",
       "01BR023                         ENSG00000144935   \n",
       "01BR023                         ENSG00000109814   \n",
       "\n",
       "Name       HGNC_UCSC_ID(supplied_by_UCSC) Oreganno_Build  \\\n",
       "Patient_ID                                                 \n",
       "01BR023                        uc003fem.5            NaN   \n",
       "01BR023                        uc001ezu.1            NaN   \n",
       "01BR023                        uc002tcb.3            NaN   \n",
       "01BR023                        uc003evb.4            NaN   \n",
       "01BR023                        uc003guk.3            NaN   \n",
       "\n",
       "Name       Simple_Uniprot_alt_uniprot_accessions  \\\n",
       "Patient_ID                                         \n",
       "01BR023                                   A8K7P8   \n",
       "01BR023                     Q01720|Q5T583|Q9UC71   \n",
       "01BR023              Q4ZG89|Q53TU0|Q96GQ4|Q9BRI8   \n",
       "01BR023                                   Q14CE4   \n",
       "01BR023                     B3KUU2|B4DN25|O60589   \n",
       "\n",
       "Name                                             dbSNP_TOPMED  \\\n",
       "Patient_ID                                                      \n",
       "01BR023                                                   NaN   \n",
       "01BR023     0.99991239806320081,0.00001592762487257,0.0000...   \n",
       "01BR023               0.99995221712538226,0.00004778287461773   \n",
       "01BR023                                                   NaN   \n",
       "01BR023                                                   NaN   \n",
       "\n",
       "Name        HGNC_Entrez_Gene_ID(supplied_by_NCBI)  COHORT  getz washu  \n",
       "Patient_ID                                                             \n",
       "01BR023                                     590.0    BRCA  True  True  \n",
       "01BR023                                    2312.0    BRCA   NaN  True  \n",
       "01BR023                                   84804.0    BRCA  True   NaN  \n",
       "01BR023                                    7220.0    BRCA  True  True  \n",
       "01BR023                                    7358.0    BRCA  True   NaN  \n",
       "\n",
       "[5 rows x 131 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Select mutations for specific patients of interest\n",
    "\n",
    "patients_of_interest = [\"01BR023\", \"11BR015\", \"01BR025\"]\n",
    "mutations_of_interest = normal_patient_mutations.loc[patients_of_interest]\n",
    "mutations_of_interest.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b6ffef8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Patient_ID\n",
       "11BR015    32\n",
       "01BR023    26\n",
       "01BR025    25\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check how many mutations each patient has\n",
    "\n",
    "mutations_of_interest.index.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "50e301e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene\n",
       "BCHE          1\n",
       "AC008770.2    1\n",
       "KRTAP4-1      1\n",
       "GLG1          1\n",
       "ZNF23         1\n",
       "             ..\n",
       "DIABLO        1\n",
       "PACSIN2       1\n",
       "SLC66A1       1\n",
       "LILRB5        1\n",
       "EDC3          1\n",
       "Name: count, Length: 83, dtype: int64"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check how many mutations per gene\n",
    "\n",
    "mutations_of_interest[\"Gene\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4613bebb",
   "metadata": {},
   "source": [
    "This shows that these three patients do not share any mutations with each other. They must be similar in some other way. We'll look at CNV next."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2f91ce3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSG00000121410.10</th>\n",
       "      <th>ENSG00000148584.13</th>\n",
       "      <th>ENSG00000175899.13</th>\n",
       "      <th>ENSG00000166535.18</th>\n",
       "      <th>ENSG00000184389.9</th>\n",
       "      <th>ENSG00000128274.14</th>\n",
       "      <th>ENSG00000118017.3</th>\n",
       "      <th>ENSG00000094914.11</th>\n",
       "      <th>ENSG00000081760.15</th>\n",
       "      <th>ENSG00000114771.12</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000086827.7</th>\n",
       "      <th>ENSG00000174442.10</th>\n",
       "      <th>ENSG00000122952.15</th>\n",
       "      <th>ENSG00000070476.13</th>\n",
       "      <th>ENSG00000203995.8</th>\n",
       "      <th>ENSG00000162378.11</th>\n",
       "      <th>ENSG00000159840.14</th>\n",
       "      <th>ENSG00000074755.13</th>\n",
       "      <th>ENSG00000036549.11</th>\n",
       "      <th>ENSG00000091436.15</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>-0.08158</td>\n",
       "      <td>0.13422</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>-0.40016</td>\n",
       "      <td>0.40444</td>\n",
       "      <td>-0.04538</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>0.00930</td>\n",
       "      <td>-0.04538</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.51938</td>\n",
       "      <td>0.04571</td>\n",
       "      <td>0.13422</td>\n",
       "      <td>0.10406</td>\n",
       "      <td>-0.40016</td>\n",
       "      <td>-0.40016</td>\n",
       "      <td>-0.14680</td>\n",
       "      <td>0.22234</td>\n",
       "      <td>0.17992</td>\n",
       "      <td>0.03808</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11BR015</th>\n",
       "      <td>0.05911</td>\n",
       "      <td>0.20725</td>\n",
       "      <td>0.09614</td>\n",
       "      <td>0.09614</td>\n",
       "      <td>0.24075</td>\n",
       "      <td>0.00250</td>\n",
       "      <td>0.08565</td>\n",
       "      <td>0.09614</td>\n",
       "      <td>0.09614</td>\n",
       "      <td>0.08565</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.58485</td>\n",
       "      <td>-0.34654</td>\n",
       "      <td>0.20725</td>\n",
       "      <td>0.15122</td>\n",
       "      <td>0.24075</td>\n",
       "      <td>0.24075</td>\n",
       "      <td>0.00264</td>\n",
       "      <td>0.16272</td>\n",
       "      <td>0.24075</td>\n",
       "      <td>-0.16318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR025</th>\n",
       "      <td>0.00226</td>\n",
       "      <td>0.01699</td>\n",
       "      <td>-0.00168</td>\n",
       "      <td>-0.00168</td>\n",
       "      <td>-0.00043</td>\n",
       "      <td>-0.00361</td>\n",
       "      <td>0.00382</td>\n",
       "      <td>-0.00168</td>\n",
       "      <td>-0.00168</td>\n",
       "      <td>0.00382</td>\n",
       "      <td>...</td>\n",
       "      <td>0.01101</td>\n",
       "      <td>-0.00746</td>\n",
       "      <td>0.01699</td>\n",
       "      <td>0.00382</td>\n",
       "      <td>-0.00043</td>\n",
       "      <td>-0.00043</td>\n",
       "      <td>0.01416</td>\n",
       "      <td>0.00070</td>\n",
       "      <td>-0.00043</td>\n",
       "      <td>0.00030</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 18926 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                      A1BG               A1CF                A2M  \\\n",
       "Database_ID ENSG00000121410.10 ENSG00000148584.13 ENSG00000175899.13   \n",
       "Patient_ID                                                             \n",
       "01BR023               -0.08158            0.13422            0.00930   \n",
       "11BR015                0.05911            0.20725            0.09614   \n",
       "01BR025                0.00226            0.01699           -0.00168   \n",
       "\n",
       "Name                     A2ML1           A3GALT2             A4GALT  \\\n",
       "Database_ID ENSG00000166535.18 ENSG00000184389.9 ENSG00000128274.14   \n",
       "Patient_ID                                                            \n",
       "01BR023                0.00930          -0.40016            0.40444   \n",
       "11BR015                0.09614           0.24075            0.00250   \n",
       "01BR025               -0.00168          -0.00043           -0.00361   \n",
       "\n",
       "Name                    A4GNT               AAAS               AACS  \\\n",
       "Database_ID ENSG00000118017.3 ENSG00000094914.11 ENSG00000081760.15   \n",
       "Patient_ID                                                            \n",
       "01BR023              -0.04538            0.00930            0.00930   \n",
       "11BR015               0.08565            0.09614            0.09614   \n",
       "01BR025               0.00382           -0.00168           -0.00168   \n",
       "\n",
       "Name                     AADAC  ...              ZW10             ZWILCH  \\\n",
       "Database_ID ENSG00000114771.12  ... ENSG00000086827.7 ENSG00000174442.10   \n",
       "Patient_ID                      ...                                        \n",
       "01BR023               -0.04538  ...          -0.51938            0.04571   \n",
       "11BR015                0.08565  ...          -0.58485           -0.34654   \n",
       "01BR025                0.00382  ...           0.01101           -0.00746   \n",
       "\n",
       "Name                     ZWINT               ZXDC            ZYG11A  \\\n",
       "Database_ID ENSG00000122952.15 ENSG00000070476.13 ENSG00000203995.8   \n",
       "Patient_ID                                                            \n",
       "01BR023                0.13422            0.10406          -0.40016   \n",
       "11BR015                0.20725            0.15122           0.24075   \n",
       "01BR025                0.01699            0.00382          -0.00043   \n",
       "\n",
       "Name                    ZYG11B                ZYX              ZZEF1  \\\n",
       "Database_ID ENSG00000162378.11 ENSG00000159840.14 ENSG00000074755.13   \n",
       "Patient_ID                                                             \n",
       "01BR023               -0.40016           -0.14680            0.22234   \n",
       "11BR015                0.24075            0.00264            0.16272   \n",
       "01BR025               -0.00043            0.01416            0.00070   \n",
       "\n",
       "Name                      ZZZ3                 pk  \n",
       "Database_ID ENSG00000036549.11 ENSG00000091436.15  \n",
       "Patient_ID                                         \n",
       "01BR023                0.17992            0.03808  \n",
       "11BR015                0.24075           -0.16318  \n",
       "01BR025               -0.00043            0.00030  \n",
       "\n",
       "[3 rows x 18926 columns]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_of_interest = candidate_cnv_patients.loc[patients_of_interest]\n",
    "cnv_of_interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "639cfcff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patient_ID   01BR023   11BR015   01BR025\n",
      "Patient_ID                              \n",
      "01BR023     1.000000  0.387652  0.047473\n",
      "11BR015     0.387652  1.000000  0.105621\n",
      "01BR025     0.047473  0.105621  1.000000\n",
      "Patient_ID   01BR023   11BR015   01BR025\n",
      "Patient_ID                              \n",
      "01BR023     1.000000  0.316389  0.137207\n",
      "11BR015     0.316389  1.000000  0.195531\n",
      "01BR025     0.137207  0.195531  1.000000\n"
     ]
    }
   ],
   "source": [
    "cnv_pearson_corr = cnv_of_interest.T.corr(method=\"pearson\")\n",
    "cnv_spearman_corr = cnv_of_interest.T.corr(method=\"spearman\")\n",
    "\n",
    "print(cnv_pearson_corr)\n",
    "print(cnv_spearman_corr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b91918ae",
   "metadata": {},
   "source": [
    "We previously had determined that there are no extreme CNV values for any of these patients. The correlation coefficients show that there is no significant correlation betweeen CNV patterns in any pair of these patients. This means something else must be common among these patients."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f93beec5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: gseapy in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (1.1.11)\n",
      "Requirement already satisfied: numpy>=1.13.0 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from gseapy) (2.3.3)\n",
      "Requirement already satisfied: scipy in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from gseapy) (1.16.3)\n",
      "Requirement already satisfied: pandas in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from gseapy) (2.3.3)\n",
      "Requirement already satisfied: matplotlib>=2.2 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from gseapy) (3.10.6)\n",
      "Requirement already satisfied: requests in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from gseapy) (2.32.5)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from matplotlib>=2.2->gseapy) (1.3.1)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from matplotlib>=2.2->gseapy) (0.11.0)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from matplotlib>=2.2->gseapy) (4.60.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from matplotlib>=2.2->gseapy) (1.4.8)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from matplotlib>=2.2->gseapy) (25.0)\n",
      "Requirement already satisfied: pillow>=8 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from matplotlib>=2.2->gseapy) (11.3.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from matplotlib>=2.2->gseapy) (3.2.0)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from matplotlib>=2.2->gseapy) (2.9.0.post0)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from python-dateutil>=2.7->matplotlib>=2.2->gseapy) (1.17.0)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from pandas->gseapy) (2025.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from pandas->gseapy) (2025.2)\n",
      "Requirement already satisfied: charset_normalizer<4,>=2 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from requests->gseapy) (3.4.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from requests->gseapy) (3.11)\n",
      "Requirement already satisfied: urllib3<3,>=1.21.1 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from requests->gseapy) (2.5.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\jason\\miniconda3\\envs\\cs312\\lib\\site-packages (from requests->gseapy) (2025.11.12)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install gseapy\n",
    "\n",
    "import gseapy as gp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9efa990",
   "metadata": {},
   "source": [
    "Our next step is to perform Gene Set Enrichment Analysis to see if these patients have problems with genes involved in the same group or pathway. We'll perform GSEA on mutation data first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "25cb41cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enrichment results for patient 01BR023:\n",
      "          Gene_set                                    Term Overlap   P-value  \\\n",
      "0  KEGG_2021_Human                  MAPK signaling pathway   3/294  0.006365   \n",
      "1  KEGG_2021_Human    Leukocyte transendothelial migration   2/114  0.009573   \n",
      "2  KEGG_2021_Human          Sphingolipid signaling pathway   2/119  0.010393   \n",
      "3  KEGG_2021_Human              Osteoclast differentiation   2/127  0.011769   \n",
      "4  KEGG_2021_Human                          Oocyte meiosis   2/129  0.012125   \n",
      "5  KEGG_2021_Human                    Dopaminergic synapse   2/132  0.012667   \n",
      "6  KEGG_2021_Human                 Cell adhesion molecules   2/148  0.015736   \n",
      "7  KEGG_2021_Human  Adrenergic signaling in cardiomyocytes   2/150  0.016140   \n",
      "8  KEGG_2021_Human                     Thiamine metabolism    1/15  0.019330   \n",
      "9  KEGG_2021_Human   Pathogenic Escherichia coli infection   2/197  0.026866   \n",
      "\n",
      "   Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0          0.169474            0                     0    8.822501   \n",
      "1          0.169474            0                     0   14.778274   \n",
      "2          0.169474            0                     0   14.143162   \n",
      "3          0.169474            0                     0   13.232667   \n",
      "4          0.169474            0                     0   13.022966   \n",
      "5          0.169474            0                     0   12.720513   \n",
      "6          0.169474            0                     0   11.317352   \n",
      "7          0.169474            0                     0   11.163288   \n",
      "8          0.180414            0                     0   57.028571   \n",
      "9          0.211586            0                     0    8.452564   \n",
      "\n",
      "   Combined Score             Genes  \n",
      "0       44.615199  MAPT;FLNC;MAPK13  \n",
      "1       68.702017       OCLN;MAPK13  \n",
      "2       64.586208    PPP2R5D;MAPK13  \n",
      "3       58.783410     LILRB5;MAPK13  \n",
      "4       57.463956    PPP2R5D;MAPK13  \n",
      "5       55.572534    PPP2R5D;MAPK13  \n",
      "6       46.987150       OCLN;LRRC4B  \n",
      "7       46.064510    PPP2R5D;MAPK13  \n",
      "8      225.039974               AK5  \n",
      "9       30.571984       OCLN;MAPK13  \n",
      "Enrichment results for patient 01BR025:\n",
      "          Gene_set                                         Term Overlap  \\\n",
      "0  KEGG_2021_Human                   cGMP-PKG signaling pathway   2/167   \n",
      "1  KEGG_2021_Human                              RNA degradation    1/79   \n",
      "2  KEGG_2021_Human             Fc gamma R-mediated phagocytosis    1/97   \n",
      "3  KEGG_2021_Human             Herpes simplex virus 1 infection   2/498   \n",
      "4  KEGG_2021_Human         Leukocyte transendothelial migration   1/114   \n",
      "5  KEGG_2021_Human                          Platelet activation   1/124   \n",
      "6  KEGG_2021_Human                     Apelin signaling pathway   1/137   \n",
      "7  KEGG_2021_Human                      Cell adhesion molecules   1/148   \n",
      "8  KEGG_2021_Human                               Tight junction   1/169   \n",
      "9  KEGG_2021_Human  Protein processing in endoplasmic reticulum   1/171   \n",
      "\n",
      "    P-value  Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0  0.018332          0.238316            0                     0   10.440053   \n",
      "1  0.094262          0.251284            0                     0   10.628739   \n",
      "2  0.114513          0.251284            0                     0    8.628038   \n",
      "3  0.127652          0.251284            0                     0    3.414972   \n",
      "4  0.133240          0.251284            0                     0    7.323746   \n",
      "5  0.144078          0.251284            0                     0    6.724932   \n",
      "6  0.157972          0.251284            0                     0    6.078125   \n",
      "7  0.169560          0.251284            0                     0    5.620181   \n",
      "8  0.191258          0.251284            0                     0    4.912450   \n",
      "9  0.193296          0.251284            0                     0    4.854167   \n",
      "\n",
      "   Combined Score          Genes  \n",
      "0       41.750899     VASP;MEF2B  \n",
      "1       25.101698           EDC3  \n",
      "2       18.697519           VASP  \n",
      "3        7.029553  ZNF780B;ZNF23  \n",
      "4       14.761759           VASP  \n",
      "5       13.028899           VASP  \n",
      "6       11.216182          MEF2B  \n",
      "7        9.973291           GLG1  \n",
      "8        8.125848           VASP  \n",
      "9        7.977991          RRBP1  \n",
      "Enrichment results for patient 11BR015:\n",
      "          Gene_set                                  Term Overlap   P-value  \\\n",
      "0  KEGG_2021_Human           Hypertrophic cardiomyopathy    2/90  0.009099   \n",
      "1  KEGG_2021_Human                Dilated cardiomyopathy    2/96  0.010298   \n",
      "2  KEGG_2021_Human        Glycerophospholipid metabolism    2/98  0.010713   \n",
      "3  KEGG_2021_Human      Vitamin digestion and absorption    1/24  0.037723   \n",
      "4  KEGG_2021_Human       alpha-Linolenic acid metabolism    1/25  0.039264   \n",
      "5  KEGG_2021_Human              Linoleic acid metabolism    1/29  0.045406   \n",
      "6  KEGG_2021_Human      Mucin type O-glycan biosynthesis    1/36  0.056064   \n",
      "7  KEGG_2021_Human          Fat digestion and absorption    1/43  0.066605   \n",
      "8  KEGG_2021_Human                      ABC transporters    1/45  0.069596   \n",
      "9  KEGG_2021_Human  Other types of O-glycan biosynthesis    1/47  0.072578   \n",
      "\n",
      "   Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \\\n",
      "0          0.164260            0                     0   15.060606   \n",
      "1          0.164260            0                     0   14.095035   \n",
      "2          0.164260            0                     0   13.800000   \n",
      "3          0.289495            0                     0   27.973352   \n",
      "4          0.289495            0                     0   26.806452   \n",
      "5          0.289495            0                     0   22.972350   \n",
      "6          0.289495            0                     0   18.371429   \n",
      "7          0.289495            0                     0   15.304147   \n",
      "8          0.289495            0                     0   14.607038   \n",
      "9          0.289495            0                     0   13.970547   \n",
      "\n",
      "   Combined Score       Genes  \n",
      "0       70.779228    RYR2;TTN  \n",
      "1       64.496093    RYR2;TTN  \n",
      "2       62.601392  PLB1;PLA1A  \n",
      "3       91.682435        PLB1  \n",
      "4       86.784309        PLB1  \n",
      "5       71.032901        PLB1  \n",
      "6       52.933045     GALNT18  \n",
      "7       41.458463       ABCA1  \n",
      "8       38.928376       ABCA1  \n",
      "9       36.646034     GALNT18  \n"
     ]
    }
   ],
   "source": [
    "# Perform gene encrichment analysis for each patient\n",
    "\n",
    "enrich_results = {}\n",
    "for patient, gene in mutations_of_interest.groupby(mutations_of_interest.index):\n",
    "    gene_list = gene[\"Gene\"].tolist()\n",
    "    enr = gp.enrichr(gene_list=gene_list,\n",
    "                     gene_sets=[\"KEGG_2021_Human\"],\n",
    "                     outdir=None)\n",
    "    enrich_results[patient] = enr.results\n",
    "\n",
    "for patient, result in enrich_results.items():\n",
    "    print(f\"Enrichment results for patient {patient}:\")\n",
    "    print(result.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f91f4df6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Endocytosis', 'Focal adhesion'}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_results = {patient: df[df['Adjusted P-value'] < 0.4] \n",
    "                    for patient, df in enrich_results.items()}\n",
    "shared_pathways = set.intersection(*(set(df['Term']) for df in filtered_results.values()))\n",
    "\n",
    "shared_pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f04b625",
   "metadata": {},
   "source": [
    "After filtering for groups using a very generous adjusted p-value of 0.4, the shared pathways between the three patients are endocytosis and focal adhesion. This means they all contain mutations affecting those functions. This could be meaningful, but the pathways aren't directly connected to cancer, and the p-values are pretty high. We'll explore further by performing GSEA on protein expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "e82bb503",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>ARF5</th>\n",
       "      <th>M6PR</th>\n",
       "      <th>ESRRA</th>\n",
       "      <th>FKBP4</th>\n",
       "      <th>NDUFAF7</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>DBNDD1</th>\n",
       "      <th>SEMA3F</th>\n",
       "      <th>CYP51A1</th>\n",
       "      <th>USP28</th>\n",
       "      <th>...</th>\n",
       "      <th>ANK2</th>\n",
       "      <th>TNK2</th>\n",
       "      <th>MYO6</th>\n",
       "      <th>EED</th>\n",
       "      <th>DDHD1</th>\n",
       "      <th>GBF1</th>\n",
       "      <th>LDB1</th>\n",
       "      <th>WIZ</th>\n",
       "      <th>RFX7</th>\n",
       "      <th>SWSAP1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSP00000000233.5</th>\n",
       "      <th>ENSP00000000412.3</th>\n",
       "      <th>ENSP00000000442.6</th>\n",
       "      <th>ENSP00000001008.4</th>\n",
       "      <th>ENSP00000002125.4</th>\n",
       "      <th>ENSP00000002165.5</th>\n",
       "      <th>ENSP00000002501.6</th>\n",
       "      <th>ENSP00000002829.3</th>\n",
       "      <th>ENSP00000003100.8</th>\n",
       "      <th>ENSP00000003302.4</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSP00000500269.1</th>\n",
       "      <th>ENSP00000500452.1</th>\n",
       "      <th>ENSP00000500710.1</th>\n",
       "      <th>ENSP00000500914.1</th>\n",
       "      <th>ENSP00000500986.2</th>\n",
       "      <th>ENSP00000501064.1</th>\n",
       "      <th>ENSP00000501277.1</th>\n",
       "      <th>ENSP00000501300.1</th>\n",
       "      <th>ENSP00000501317.1</th>\n",
       "      <th>ENSP00000501355.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>-0.303983</td>\n",
       "      <td>-0.114765</td>\n",
       "      <td>-0.765492</td>\n",
       "      <td>0.469224</td>\n",
       "      <td>0.333034</td>\n",
       "      <td>-0.328081</td>\n",
       "      <td>-0.472246</td>\n",
       "      <td>0.409787</td>\n",
       "      <td>-0.230430</td>\n",
       "      <td>-0.661301</td>\n",
       "      <td>...</td>\n",
       "      <td>0.372218</td>\n",
       "      <td>0.572301</td>\n",
       "      <td>-0.611428</td>\n",
       "      <td>0.078793</td>\n",
       "      <td>-0.320561</td>\n",
       "      <td>-0.129917</td>\n",
       "      <td>0.082966</td>\n",
       "      <td>0.342258</td>\n",
       "      <td>0.183923</td>\n",
       "      <td>-0.727190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11BR015</th>\n",
       "      <td>0.131767</td>\n",
       "      <td>-0.208966</td>\n",
       "      <td>-0.206715</td>\n",
       "      <td>1.171356</td>\n",
       "      <td>0.054821</td>\n",
       "      <td>-0.526126</td>\n",
       "      <td>-0.350226</td>\n",
       "      <td>0.826188</td>\n",
       "      <td>-0.687686</td>\n",
       "      <td>-0.589347</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.311004</td>\n",
       "      <td>-0.243446</td>\n",
       "      <td>-0.718078</td>\n",
       "      <td>-0.374621</td>\n",
       "      <td>1.124724</td>\n",
       "      <td>-0.101972</td>\n",
       "      <td>0.601303</td>\n",
       "      <td>0.842491</td>\n",
       "      <td>0.107564</td>\n",
       "      <td>0.754653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR025</th>\n",
       "      <td>0.514449</td>\n",
       "      <td>-0.181170</td>\n",
       "      <td>-0.273682</td>\n",
       "      <td>0.526654</td>\n",
       "      <td>0.422711</td>\n",
       "      <td>0.675484</td>\n",
       "      <td>0.032640</td>\n",
       "      <td>-0.006112</td>\n",
       "      <td>-0.418856</td>\n",
       "      <td>0.448302</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.146638</td>\n",
       "      <td>0.189738</td>\n",
       "      <td>0.143171</td>\n",
       "      <td>-0.665344</td>\n",
       "      <td>-1.401882</td>\n",
       "      <td>0.237409</td>\n",
       "      <td>-0.359958</td>\n",
       "      <td>0.109089</td>\n",
       "      <td>0.039473</td>\n",
       "      <td>1.250211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 9538 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                     ARF5              M6PR             ESRRA  \\\n",
       "Database_ID ENSP00000000233.5 ENSP00000000412.3 ENSP00000000442.6   \n",
       "Patient_ID                                                          \n",
       "01BR023             -0.303983         -0.114765         -0.765492   \n",
       "11BR015              0.131767         -0.208966         -0.206715   \n",
       "01BR025              0.514449         -0.181170         -0.273682   \n",
       "\n",
       "Name                    FKBP4           NDUFAF7             FUCA2  \\\n",
       "Database_ID ENSP00000001008.4 ENSP00000002125.4 ENSP00000002165.5   \n",
       "Patient_ID                                                          \n",
       "01BR023              0.469224          0.333034         -0.328081   \n",
       "11BR015              1.171356          0.054821         -0.526126   \n",
       "01BR025              0.526654          0.422711          0.675484   \n",
       "\n",
       "Name                   DBNDD1            SEMA3F           CYP51A1  \\\n",
       "Database_ID ENSP00000002501.6 ENSP00000002829.3 ENSP00000003100.8   \n",
       "Patient_ID                                                          \n",
       "01BR023             -0.472246          0.409787         -0.230430   \n",
       "11BR015             -0.350226          0.826188         -0.687686   \n",
       "01BR025              0.032640         -0.006112         -0.418856   \n",
       "\n",
       "Name                    USP28  ...              ANK2              TNK2  \\\n",
       "Database_ID ENSP00000003302.4  ... ENSP00000500269.1 ENSP00000500452.1   \n",
       "Patient_ID                     ...                                       \n",
       "01BR023             -0.661301  ...          0.372218          0.572301   \n",
       "11BR015             -0.589347  ...         -0.311004         -0.243446   \n",
       "01BR025              0.448302  ...         -1.146638          0.189738   \n",
       "\n",
       "Name                     MYO6               EED             DDHD1  \\\n",
       "Database_ID ENSP00000500710.1 ENSP00000500914.1 ENSP00000500986.2   \n",
       "Patient_ID                                                          \n",
       "01BR023             -0.611428          0.078793         -0.320561   \n",
       "11BR015             -0.718078         -0.374621          1.124724   \n",
       "01BR025              0.143171         -0.665344         -1.401882   \n",
       "\n",
       "Name                     GBF1              LDB1               WIZ  \\\n",
       "Database_ID ENSP00000501064.1 ENSP00000501277.1 ENSP00000501300.1   \n",
       "Patient_ID                                                          \n",
       "01BR023             -0.129917          0.082966          0.342258   \n",
       "11BR015             -0.101972          0.601303          0.842491   \n",
       "01BR025              0.237409         -0.359958          0.109089   \n",
       "\n",
       "Name                     RFX7            SWSAP1  \n",
       "Database_ID ENSP00000501317.1 ENSP00000501355.1  \n",
       "Patient_ID                                       \n",
       "01BR023              0.183923         -0.727190  \n",
       "11BR015              0.107564          0.754653  \n",
       "01BR025              0.039473          1.250211  \n",
       "\n",
       "[3 rows x 9538 columns]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Analyze protein expression for patients of interest (and filter out columns with na values)\n",
    "\n",
    "proteins_of_interest = normal_prot.loc[patients_of_interest]\n",
    "proteins_of_interest = proteins_of_interest.dropna(axis=1)\n",
    "proteins_of_interest.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "41a31239",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get proteins where all patient values are extreme\n",
    "\n",
    "extreme_proteins = proteins_of_interest.columns[(abs(proteins_of_interest) > 1.5).all(axis=0)]\n",
    "len(extreme_proteins)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e17eb8db",
   "metadata": {},
   "source": [
    "First, we filtered to only include proteins that are expressed at extreme levels (< -1.5 or greater than 1.5) in all three patients. This left us with 50 proteins to perform GSEA on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "3da841eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Antimicrobial Humoral Response (GO:0019730)</td>\n",
       "      <td>0.035740</td>\n",
       "      <td>4/100</td>\n",
       "      <td>CXCL9;HLA-A;S100A7;KRT6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Defense Response To Bacterium (GO:0042742)</td>\n",
       "      <td>0.035740</td>\n",
       "      <td>5/204</td>\n",
       "      <td>CHGA;HLA-A;GBP1;S100A8;KRT6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Negative Regulation Of Growth (GO:0045926)</td>\n",
       "      <td>0.040387</td>\n",
       "      <td>4/124</td>\n",
       "      <td>MT1G;TP53;ALOX15B;MT1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Antimicrobial Humoral Immune Response Mediated...</td>\n",
       "      <td>0.047297</td>\n",
       "      <td>3/65</td>\n",
       "      <td>CXCL9;S100A7;KRT6A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Cellular Response To Zinc Ion (GO:0071294)</td>\n",
       "      <td>0.047297</td>\n",
       "      <td>2/15</td>\n",
       "      <td>MT1G;MT1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Prostanoid Metabolic Process (GO:0006692)</td>\n",
       "      <td>0.047297</td>\n",
       "      <td>2/16</td>\n",
       "      <td>HPGD;PTGES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Positive Regulation Of Macrophage Derived Foam...</td>\n",
       "      <td>0.047297</td>\n",
       "      <td>2/17</td>\n",
       "      <td>AGTR1;ALOX15B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Positive Regulation Of G Protein-Coupled Recep...</td>\n",
       "      <td>0.047297</td>\n",
       "      <td>2/18</td>\n",
       "      <td>CHGA;KLK6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Negative Regulation Of T Cell Receptor Signali...</td>\n",
       "      <td>0.047297</td>\n",
       "      <td>2/18</td>\n",
       "      <td>SH2D1A;GBP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Cellular Response To Copper Ion (GO:0071280)</td>\n",
       "      <td>0.047536</td>\n",
       "      <td>2/20</td>\n",
       "      <td>MT1G;MT1E</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Term  Adjusted P-value  \\\n",
       "0        Antimicrobial Humoral Response (GO:0019730)          0.035740   \n",
       "1         Defense Response To Bacterium (GO:0042742)          0.035740   \n",
       "2         Negative Regulation Of Growth (GO:0045926)          0.040387   \n",
       "3  Antimicrobial Humoral Immune Response Mediated...          0.047297   \n",
       "4         Cellular Response To Zinc Ion (GO:0071294)          0.047297   \n",
       "5          Prostanoid Metabolic Process (GO:0006692)          0.047297   \n",
       "6  Positive Regulation Of Macrophage Derived Foam...          0.047297   \n",
       "7  Positive Regulation Of G Protein-Coupled Recep...          0.047297   \n",
       "8  Negative Regulation Of T Cell Receptor Signali...          0.047297   \n",
       "9       Cellular Response To Copper Ion (GO:0071280)          0.047536   \n",
       "\n",
       "  Overlap                         Genes  \n",
       "0   4/100      CXCL9;HLA-A;S100A7;KRT6A  \n",
       "1   5/204  CHGA;HLA-A;GBP1;S100A8;KRT6A  \n",
       "2   4/124        MT1G;TP53;ALOX15B;MT1E  \n",
       "3    3/65            CXCL9;S100A7;KRT6A  \n",
       "4    2/15                     MT1G;MT1E  \n",
       "5    2/16                    HPGD;PTGES  \n",
       "6    2/17                 AGTR1;ALOX15B  \n",
       "7    2/18                     CHGA;KLK6  \n",
       "8    2/18                   SH2D1A;GBP1  \n",
       "9    2/20                     MT1G;MT1E  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_list = extreme_proteins.get_level_values(0).tolist()\n",
    "\n",
    "enr_proteins = gp.enrichr(\n",
    "    gene_list=protein_list,\n",
    "    gene_sets=[\"KEGG_2021_Human\", \"GO_Biological_Process_2023\"],\n",
    "    outdir=None\n",
    ")\n",
    "\n",
    "significant = enr_proteins.results[enr_proteins.results['Adjusted P-value'] < 0.05]\n",
    "significant[['Term', 'Adjusted P-value', 'Overlap', 'Genes']].head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d248e3ae",
   "metadata": {},
   "source": [
    "One pathway stands out here: negative regulation of growth. It has 4 genes involved that are shared between the three patients, and a significant adjusted p-value of 0.04. We'll take a closer look at these four genes next."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b5ad2e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th colspan=\"2\" halign=\"left\">MT1G</th>\n",
       "      <th>TP53</th>\n",
       "      <th>ALOX15B</th>\n",
       "      <th>MT1E</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSP00000369139.4</th>\n",
       "      <th>ENSP00000391397.2</th>\n",
       "      <th>ENSP00000269305.4</th>\n",
       "      <th>ENSP00000369530.4</th>\n",
       "      <th>ENSP00000307706.5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>-2.342940</td>\n",
       "      <td>-2.335595</td>\n",
       "      <td>-1.668498</td>\n",
       "      <td>-2.281744</td>\n",
       "      <td>-2.866936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11BR015</th>\n",
       "      <td>-2.219948</td>\n",
       "      <td>-1.426507</td>\n",
       "      <td>-2.005719</td>\n",
       "      <td>-2.829229</td>\n",
       "      <td>-1.998386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR025</th>\n",
       "      <td>-3.890109</td>\n",
       "      <td>0.112658</td>\n",
       "      <td>-1.618140</td>\n",
       "      <td>-1.833970</td>\n",
       "      <td>-2.135432</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                     MT1G                                TP53  \\\n",
       "Database_ID ENSP00000369139.4 ENSP00000391397.2 ENSP00000269305.4   \n",
       "Patient_ID                                                          \n",
       "01BR023             -2.342940         -2.335595         -1.668498   \n",
       "11BR015             -2.219948         -1.426507         -2.005719   \n",
       "01BR025             -3.890109          0.112658         -1.618140   \n",
       "\n",
       "Name                  ALOX15B              MT1E  \n",
       "Database_ID ENSP00000369530.4 ENSP00000307706.5  \n",
       "Patient_ID                                       \n",
       "01BR023             -2.281744         -2.866936  \n",
       "11BR015             -2.829229         -1.998386  \n",
       "01BR025             -1.833970         -2.135432  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check expression levels of specific proteins related to negative growth regulation found previously\n",
    "\n",
    "negative_growth_proteins = [\"MT1G\", \"TP53\", \"ALOX15B\", \"MT1E\"]\n",
    "proteins_of_interest[negative_growth_proteins]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1e1e8be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>...</th>\n",
       "      <th>HGNC_UniProt_ID(supplied_by_UniProt)</th>\n",
       "      <th>HGNC_Ensembl_ID(supplied_by_Ensembl)</th>\n",
       "      <th>HGNC_UCSC_ID(supplied_by_UCSC)</th>\n",
       "      <th>Oreganno_Build</th>\n",
       "      <th>Simple_Uniprot_alt_uniprot_accessions</th>\n",
       "      <th>dbSNP_TOPMED</th>\n",
       "      <th>HGNC_Entrez_Gene_ID(supplied_by_NCBI)</th>\n",
       "      <th>COHORT</th>\n",
       "      <th>getz</th>\n",
       "      <th>washu</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 131 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Gene, Mutation, Location, Entrez_Gene_Id, NCBI_Build, Chromosome, Start_Position, End_Position, Strand, Variant_Type, Reference_Allele, Tumor_Seq_Allele1, Tumor_Seq_Allele2, Tumor_Sample_Barcode, Matched_Norm_Sample_Barcode, Validation_Method, Score, BAM_File, Tumor_Sample_UUID, Matched_Norm_Sample_UUID, Genome_Change, Annotation_Transcript, Transcript_Strand, Transcript_Exon, Transcript_Position, cDNA_Change, Codon_Change, Other_Transcripts, Refseq_mRNA_Id, Refseq_prot_Id, SwissProt_acc_Id, SwissProt_entry_Id, Description, GO_Biological_Process, GO_Cellular_Component, GO_Molecular_Function, COSMIC_overlapping_mutations, COSMIC_fusion_genes, COSMIC_tissue_types_affected, COSMIC_total_alterations_in_gene, DrugBank, ref_context, gc_content, CCLE_ONCOMAP_total_mutations_in_gene, CGC_Mutation_Type, CGC_Translocation_Partner, CGC_Tumor_Types_Somatic, CGC_Tumor_Types_Germline, CGC_Other_Diseases, DNARepairGenes_Activity_linked_to_OMIM, FamilialCancerDatabase_Syndromes, OREGANNO_ID, OREGANNO_Values, t_alt_count, t_ref_count, n_alt_count, n_ref_count, Gencode_34_secondaryVariantClassification, Achilles_Top_Genes, CGC_Name, CGC_GeneID, CGC_Chr, CGC_Chr_Band, CGC_Cancer_Somatic_Mut, CGC_Cancer_Germline_Mut, CGC_Cancer_Syndrome, CGC_Tissue_Type, CGC_Cancer_Molecular_Genetics, CGC_Other_Germline_Mut, ClinVar_VCF_AF_ESP, ClinVar_VCF_AF_EXAC, ClinVar_VCF_AF_TGP, ClinVar_VCF_ALLELEID, ClinVar_VCF_CLNDISDB, ClinVar_VCF_CLNDISDBINCL, ClinVar_VCF_CLNDN, ClinVar_VCF_CLNDNINCL, ClinVar_VCF_CLNHGVS, ClinVar_VCF_CLNREVSTAT, ClinVar_VCF_CLNSIG, ClinVar_VCF_CLNSIGCONF, ClinVar_VCF_CLNSIGINCL, ClinVar_VCF_CLNVC, ClinVar_VCF_CLNVCSO, ClinVar_VCF_CLNVI, ClinVar_VCF_DBVARID, ClinVar_VCF_GENEINFO, ClinVar_VCF_MC, ClinVar_VCF_ORIGIN, ClinVar_VCF_RS, ClinVar_VCF_SSR, ClinVar_VCF_ID, ClinVar_VCF_FILTER, DNARepairGenes_Chromosome_location_linked_to_NCBI_MapView, DNARepairGenes_Accession_number_linked_to_NCBI_Entrez, Familial_Cancer_Genes_Synonym, Familial_Cancer_Genes_Reference, Gencode_XHGNC_hgnc_id, HGNC_HGNC_ID, HGNC_Status, ...]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 131 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Confirming none of these proteins have been mutated\n",
    "\n",
    "mutations_of_interest[mutations_of_interest[\"Gene\"].isin(negative_growth_proteins)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "930b6b05",
   "metadata": {},
   "source": [
    "This confirms that none of these patients have mutations in any of these proteins even though they are underexpressed. We'll look at CNV next."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "ac067f9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th>MT1G</th>\n",
       "      <th>TP53</th>\n",
       "      <th>ALOX15B</th>\n",
       "      <th>MT1E</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Database_ID</th>\n",
       "      <th>ENSG00000125144.12</th>\n",
       "      <th>ENSG00000141510.14</th>\n",
       "      <th>ENSG00000179593.14</th>\n",
       "      <th>ENSG00000169715.13</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01BR023</th>\n",
       "      <td>0.06305</td>\n",
       "      <td>0.22234</td>\n",
       "      <td>0.22234</td>\n",
       "      <td>0.06305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11BR015</th>\n",
       "      <td>-0.56290</td>\n",
       "      <td>0.16272</td>\n",
       "      <td>0.16272</td>\n",
       "      <td>-0.56290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01BR025</th>\n",
       "      <td>0.01112</td>\n",
       "      <td>0.00070</td>\n",
       "      <td>0.00070</td>\n",
       "      <td>0.01112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                      MT1G               TP53            ALOX15B  \\\n",
       "Database_ID ENSG00000125144.12 ENSG00000141510.14 ENSG00000179593.14   \n",
       "Patient_ID                                                             \n",
       "01BR023                0.06305            0.22234            0.22234   \n",
       "11BR015               -0.56290            0.16272            0.16272   \n",
       "01BR025                0.01112            0.00070            0.00070   \n",
       "\n",
       "Name                      MT1E  \n",
       "Database_ID ENSG00000169715.13  \n",
       "Patient_ID                      \n",
       "01BR023                0.06305  \n",
       "11BR015               -0.56290  \n",
       "01BR025                0.01112  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Confirming CNV status of these genes\n",
    "\n",
    "cnv_of_interest[negative_growth_proteins]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e370f9a8",
   "metadata": {},
   "source": [
    "For the most part, the CNV values for these proteins are normal. While MT1G and MT1E have low CNV values (< -0.3) in one patient, there is no trend across all three patients, despite them being significantly underexpressed in all three. This suggests that something else is responsible for the underexpression of these proteins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "1116cb73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MT1G is not a known driver mutation.\n",
      "TP53 is a known driver mutation.\n",
      "ALOX15B is not a known driver mutation.\n",
      "MT1E is not a known driver mutation.\n"
     ]
    }
   ],
   "source": [
    "driver_mutations_df = pd.read_csv(\"breast_driver_mutations.tsv\", sep=\"\\t\")\n",
    "\n",
    "for protein in negative_growth_proteins:\n",
    "    if protein in driver_mutations_df[\"Symbol\"].values:\n",
    "        print(f\"{protein} is a known driver mutation.\")\n",
    "    else:\n",
    "        print(f\"{protein} is not a known driver mutation.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3deffb7d",
   "metadata": {},
   "source": [
    "Three out of four of these proteins are not known driver mutations. We have found that these proteins are significantly underexpressed in three closely related cases of breast cancer, and are associated with negative regulation of growth. These four proteins are not mutated and do not have significantly altered CNVs in any of the four patients. The underexpression of these proteins involved in negative regulation of growth is likely related to these specific cases of cancer, though we still don't know exactly why they are being underexpressed."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cs312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
