{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cbe9f645",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cptac\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "edd8c392",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Brca in module cptac.cancers.brca object:\n",
      "\n",
      "class Brca(cptac.cancers.cancer.Cancer)\n",
      " |  Brca(no_internet=False)\n",
      " |  \n",
      " |  Manages BRCA (Breast Cancer) data from various sources.\n",
      " |  \n",
      " |  This class extends the base Cancer class and initializes the BRCA data from \n",
      " |  a variety of sources including BCM, Broad Institute, MSSM, University of Michigan, \n",
      " |  Washington University, and a Harmonized dataset. \n",
      " |  \n",
      " |  Attributes:\n",
      " |      _sources (dict): A dictionary holding data from different sources.\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      Brca\n",
      " |      cptac.cancers.cancer.Cancer\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __init__(self, no_internet=False)\n",
      " |      Initializes the Brca object.\n",
      " |      \n",
      " |      Args:\n",
      " |          no_internet (bool): If True, the object will not attempt to download data from the internet. \n",
      " |                              Default is False.\n",
      " |      \n",
      " |      Raises:\n",
      " |          ValueError: If the 'no_internet' argument is not of boolean type.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from cptac.cancers.cancer.Cancer:\n",
      " |  \n",
      " |  define(self, term: str)\n",
      " |      Print the definition of a term, if it is in the dataset's list of definitions.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      term (str): term to be defined\n",
      " |      \n",
      " |      Returns: None\n",
      " |  \n",
      " |  delete_df(self, df_type: str, source: str = 'all')\n",
      " |      This function enables users to delete dataframes they no longer need to free up RAM\n",
      " |      \n",
      " |      Parameters:\n",
      " |      df_type (string): The type of dataframe to delete. For example, if the proteomics dataframe\n",
      " |          was loaded with get_proteomics() then they would pass 'proteomics' to this function to delete it.\n",
      " |      source (string): The source from which to delete the dataframe of type df_type\n",
      " |          Default: 'all' loops through all sources, checks for that df_type and deletes it if present\n",
      " |  \n",
      " |  get_CNV(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get a CNV dataframe from the specified data source.\n",
      " |  \n",
      " |  get_acetylproteomics(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get the acetylproteomics dataframe from the specified data source.\n",
      " |  \n",
      " |  get_ancestry_prediction(self, source=None, tissue_type='both', imputed=False)\n",
      " |      Get ancestry_prediction data from the specified data source\n",
      " |  \n",
      " |  get_cancer_type(self) -> str\n",
      " |      Return the cancer type for this dataset, as a string.\n",
      " |  \n",
      " |  get_cibersort(self, source=None, tissue_type='both', imputed=False)\n",
      " |      Get cibersort data from the specified data source\n",
      " |  \n",
      " |  get_circular_RNA(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get a circular RNA dataframe from the specified data source.\n",
      " |  \n",
      " |  get_clinical(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get the clinical dataframe from the specified data source.\n",
      " |  \n",
      " |  get_data_list(self) -> list\n",
      " |      Return a list of all data currently loaded into memory\n",
      " |  \n",
      " |  get_dataframe(self, data_type: str, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Checks if a given dataframe from a specific source exists and return a copy if it does.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      data_type (str): The data type of the desired dataframe.\n",
      " |      source (str, optional): The source of the dataframe. If not specified and there's only one source for the data_type, that source will be used. If not specified and there's more than one source for the data_type, DataSourceNotFoundError will be raised. Defaults to None.\n",
      " |      tissue_type (str, optional): Desired tissue type in the dataframe. Acceptable values are [\"tumor\", \"normal\", \"both\"]. Defaults to \"both\".\n",
      " |      imputed (bool, optional): Specifies whether the data is imputed. Defaults to False.\n",
      " |      \n",
      " |      Raises:\n",
      " |      DataFrameNotIncludedError: If the data_type is not included in the dataset.\n",
      " |      DataSourceNotFoundError: If the source is not found in the dataset or if no source is specified and more than one source provides the data_type.\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas.DataFrame: A dataframe containing the requested data.\n",
      " |  \n",
      " |  get_derived_molecular(self, type, source=None, tissue_type='both', imputed=False)\n",
      " |      Get type of derived_molecular dataframe. Valid types are cibersort, xcell, hla_typing, and ancestry_prediction\n",
      " |  \n",
      " |  get_docs(self, data: str, source: str = None, tissue_type: str = 'both', imputed: bool = False)\n",
      " |      Get the readme docs for the specified data type and source.\n",
      " |  \n",
      " |  get_followup(self, source=None, tissue_type='both', imputed=False)\n",
      " |      Get the followup dataframe from the specified data source.\n",
      " |  \n",
      " |  get_genotype_all_vars(self, mutations_gene: str, omics_source: str, mutations_source: str, mutations_filter: list = None, show_location: bool = True, mutation_hotspot: str = None) -> pandas.core.frame.DataFrame\n",
      " |      Return a dataframe that has the mutation type and whether or not it is a multiple mutation\n",
      " |      Parameters:\n",
      " |      mutation_genes (str, or list or array-like of str): The gene(s) to get mutation data for.\n",
      " |      mutations_filter (list, optional):  List of mutations to prioritize when filtering out multiple mutations, in order of priority.\n",
      " |      omics_source(str): Source of omics data ex \"bcm\", \"washu\", \"broad\", \"umich\"\n",
      " |      mutations_source(str): Source of mutations data ex \"bcm\", \"washu\", \"broad\", \"umich\"\n",
      " |      show_location (bool, optional): Whether to include the Location column from the mutation dataframe. Defaults to True.\n",
      " |      mutation_hotspot (optional): a list of hotspot locations\n",
      " |      \n",
      " |      Returns:\n",
      " |      A dataframe specifying the primary mutation type (as specified by mutations_filter), its location (optional), and whether there is one or multiple mutations\n",
      " |  \n",
      " |  get_hla_typing(self, source=None, tissue_type='both', imputed=False)\n",
      " |      Get hla_typing data from the specified data source\n",
      " |  \n",
      " |  get_medical_history(self, source=None, tissue_type='both')\n",
      " |      Get the medical_history dataframe.\n",
      " |  \n",
      " |  get_miRNA(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get miRNA dataframe from the specified data source.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      source (str): Select data generated by a certain institution. Available sources are ['washu']. Defaults to 'washu'.\n",
      " |  \n",
      " |  get_phosphoproteomics(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get the phosphoproteomics dataframe from the specified data source.\n",
      " |  \n",
      " |  get_proteomics(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get the proteomics dataframe from the specified data source.\n",
      " |  \n",
      " |  get_somatic_mutation(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get the somatic mutation dataframe from the specified data source.\n",
      " |  \n",
      " |  get_somatic_mutation_binary(self, source: str = None) -> pandas.core.frame.DataFrame\n",
      " |      Get the somatic_mutation_binary dataframe, which has a binary value indicating, for each location on each gene, whether there was a mutation in that gene at that location, for each sample.\n",
      " |  \n",
      " |  get_targeted_phosphoproteomics(self, source: str = None, tissue_type: str = 'both') -> pandas.core.frame.DataFrame\n",
      " |      Get the targeted_phosphoproteomics dataframe.\n",
      " |  \n",
      " |  get_targeted_proteomics(self, source: str = None, tissue_type: str = 'both') -> pandas.core.frame.DataFrame\n",
      " |      Get the targeted_proteomics dataframe.\n",
      " |  \n",
      " |  get_transcriptomics(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get the transcriptomics dataframe from the specified data source.\n",
      " |  \n",
      " |  get_tumor_purity(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
      " |      Get the tumor purity dataframe from the specified data source.\n",
      " |  \n",
      " |  get_xcell(self, source=None, tissue_type='both', imputed=False)\n",
      " |      Get xcell data from the specified data source\n",
      " |  \n",
      " |  how_to_cite(self, cancer_type: str = '', pmid: str = '', unpublished: bool = False)\n",
      " |      Print instructions for citing the data.\n",
      " |  \n",
      " |  join_metadata_to_metadata(self, df1_name: str, df2_name: str, df1_source: str = None, df2_source: str = None, cols1: str = None, cols2: str = None, how: str = 'outer', quiet: bool = False, tissue_type: str = 'both') -> pandas.core.frame.DataFrame\n",
      " |      Take specified column(s) from one metadata dataframe, and join to specified columns(s) from another metadata dataframe. Intersection (inner join) of indices is used.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      df1_name (str): Name of first metadata dataframe to select columns from.\n",
      " |      df2_name (str): Name of second metadata dataframe to select columns from.\n",
      " |      df1_source (str, optional): Name of source for the first dataframe.\n",
      " |      df2_source (str, optional): Name of source for the second dataframe.\n",
      " |      cols1 (str, or list or array-like of str, optional): Column(s) to select from df1_name.\n",
      " |          str if one key, list or array-like of str if multiple.\n",
      " |          Default of None will select entire dataframe.\n",
      " |      cols2 (str, or list or array-like of str, optional): Column(s) to select from df2_name.\n",
      " |          str if one key, list or array-like of str if multiple.\n",
      " |          Default of None will select entire dataframe.\n",
      " |      how (str, optional): How to perform the join, acceptable values are from ['outer', 'inner', 'left', 'right']. Defaults to 'outer'.\n",
      " |      quiet (bool, optional): Whether to warn when inserting NaNs. Defaults to False.\n",
      " |      tissue_type (str): Acceptable values in [\"tumor\",\"normal\",\"both\"]. Specifies the desired tissue type for the dataframe. Defaults to \"both\".\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas.DataFrame: The selected columns from the two metadata dataframes, joined into one dataframe.\n",
      " |  \n",
      " |  join_metadata_to_mutations(self, metadata_name: str, mutations_genes: str, metadata_source: str = None, mutations_source: str = None, metadata_cols: str = None, mutations_filter: str = None, show_location: bool = True, how: str = 'outer', quiet: bool = False, tissue_type: str = 'both') -> pandas.core.frame.DataFrame\n",
      " |      Select all mutations for specified gene(s), and joins them to all or part of the given metadata dataframe. Intersection (inner join) of indices is used. Each location or mutation cell contains a list, which contains the one or more location or mutation values corresponding to that sample for that gene, or a value indicating that the sample didn't have a mutation in that gene.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      metadata_name (str): Name of metadata dataframe to join the mutation data to.\n",
      " |      mutations_genes (str, or list or array-like of str): The gene(s) to get mutation data for. str if one gene, list or array-like of str if multiple.\n",
      " |      metadata_source (str, optional): Name of source for metadata.\n",
      " |      mutations_source (str, optional): Name of source for mutations data.\n",
      " |      metadata_cols (str, or list or array-like of str, optional): Gene(s) to select from the metadata dataframe. str if one gene, list or array-like of str if multiple.\n",
      " |          Default will select entire dataframe.\n",
      " |      mutations_filter (list, optional): List of mutations to prioritize when filtering out multiple mutations, in order of priority.\n",
      " |          If none of the multiple mutations in a sample are included in mutations_filter, the function will automatically prioritize truncation over missense mutations, and then mutations earlier in the sequence over later mutations.\n",
      " |          Passing an empty list will cause this default hierarchy to be applied to all samples.\n",
      " |          Default parameter of None will cause no filtering to be done, and all mutation data will be included, in a list.\n",
      " |      show_location (bool, optional): Whether to include the Location column from the mutation dataframe. Defaults to True.\n",
      " |      how (str, optional): How to perform the join, acceptable values are from ['outer', 'inner', 'left', 'right']. Defaults to 'outer'.\n",
      " |      quiet (bool, optional): Whether to warn when inserting NaNs. Defaults to False.\n",
      " |      tissue_type (str): Acceptable values in [\"tumor\",\"normal\",\"both\"]. Specifies the desired tissue type for the dataframe. Defaults to \"both\".\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas.DataFrame: The mutations for the specified gene, joined to all or part of the metadata dataframe. Each location or mutation cell contains a list, which contains the one or more location or mutation values corresponding to that sample for that gene, or a value indicating that the sample didn't have a mutation in that gene.\n",
      " |  \n",
      " |  join_metadata_to_omics(self, metadata_name: str, omics_name: str, omics_source: str = None, metadata_source: str = None, metadata_cols: str = None, omics_genes: str = None, how: str = 'outer', quiet: bool = False, tissue_type: str = 'both') -> pandas.core.frame.DataFrame\n",
      " |      Joins columns from a metadata dataframe (clinical, derived_molecular, or experimental_design) to part or all of an omics dataframe. Intersection (inner join) of indices is used.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      metadata_name (str): Name of metadata dataframe to select columns from.\n",
      " |      omics_name (str): Name of omics dataframe to join the metadata columns to.\n",
      " |      metadata_cols (str, or list or array-like of str, optional): Column(s) to select from the metadata dataframe.\n",
      " |          str if one gene, list or array-like of str if multiple.\n",
      " |          Default is None, which will select the entire metadata dataframe.\n",
      " |      omics_genes (str, or list or array-like of str, optional): Gene(s) to select data for from the omics dataframe.\n",
      " |          str if one gene, list or array-like of str if multiple.\n",
      " |          Default is None, which will select entire dataframe.\n",
      " |      how (str, optional): How to perform the join, acceptable values are from ['outer', 'inner', 'left', 'right']. Defaults to 'outer'.\n",
      " |      quiet (bool, optional): Whether to warn when inserting NaNs. Defaults to False.\n",
      " |      tissue_type (str): Acceptable values in [\"tumor\",\"normal\",\"both\"]. Specifies the desired tissue type for the dataframe. Defaults to \"both\".\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas.DataFrame: The selected metadata columns, joined with all or part of the omics dataframe.\n",
      " |  \n",
      " |  join_omics_to_mutations(self, omics_name: str, mutations_genes: str, omics_source: str, mutations_source: str, omics_genes: str = None, mutations_filter: str = None, show_location: bool = True, how: str = 'outer', quiet: bool = False, tissue_type: str = 'both', mutation_cols: str = ['Mutation', 'Location']) -> pandas.core.frame.DataFrame\n",
      " |      Joins specified mutations for a gene(s) to a selected omics dataframe using intersection (inner join) of indices.\n",
      " |      \n",
      " |      The function returns a DataFrame where each location or mutation cell contains a list. This list contains one or more location or mutation values for that sample for the gene in question. If a sample doesn't have a mutation in that gene, the cell will indicate so.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      - omics_name (str): Name of the omics dataframe to join the mutation data to.\n",
      " |      - mutations_genes (str or list[str]): The gene(s) for which to retrieve mutation data. Pass a str for one gene, or a list of str for multiple genes.\n",
      " |      - omics_source (str): Name of source for the omics data.\n",
      " |      - mutations_source (str): Name of source for the mutations data.\n",
      " |      - omics_genes (str or list[str], optional): Gene(s) to extract from the omics dataframe. Pass a str for one gene, or a list of str for multiple. Default is None, which selects the entire dataframe.\n",
      " |      - mutations_filter (list, optional): List of mutations to prioritize when filtering multiple mutations, in order of priority. Default is None, including all mutation data. Refer to full function docstring for more details.\n",
      " |      - show_location (bool, optional): Whether to include the Location column from the mutation dataframe. Default is True.\n",
      " |      - how (str, optional): Type of join operation to perform. Accepts 'outer', 'inner', 'left', 'right'. Default is 'outer'.\n",
      " |      - quiet (bool, optional): Whether to suppress warnings when inserting NaNs. Default is False.\n",
      " |      - tissue_type (str, optional): Desired tissue type for the dataframe. Accepts 'tumor', 'normal', 'both'. Default is 'both'.\n",
      " |      \n",
      " |      Returns:\n",
      " |      - pandas.DataFrame: A DataFrame of mutations for the specified gene(s), joined with the selected portion of the omics dataframe. Each cell in the location or mutation column contains a list with location or mutation values for that sample, or an indicator if the sample didn't have a mutation in that gene.\n",
      " |      \n",
      " |      Example:\n",
      " |      ```python\n",
      " |      join_omics_to_mutations(\n",
      " |          omics_name='proteomics',\n",
      " |          mutations_genes='TP53',\n",
      " |          omics_source='TCGA',\n",
      " |          mutations_source='TCGA',\n",
      " |          omics_genes='TP53',\n",
      " |          mutations_filter=['nonsynonymous SNV'],\n",
      " |          show_location=True,\n",
      " |          how='outer',\n",
      " |          quiet=True,\n",
      " |          tissue_type='tumor'\n",
      " |      )\n",
      " |      ```\n",
      " |  \n",
      " |  join_omics_to_omics(self, df1_name: str, df2_name: str, df1_source: str = None, df2_source: str = None, genes1: str = None, genes2: str = None, how: str = 'outer', quiet: bool = False, tissue_type: str = 'both') -> pandas.core.frame.DataFrame\n",
      " |      Take specified column(s) from one omics dataframe, and join to specified columns(s) from another omics dataframe. Intersection (inner join) of indices is used.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      df1_name (str): Name of first omics dataframe to select columns from.\n",
      " |      df2_name (str): Name of second omics dataframe to select columns from.\n",
      " |      df1_source (str): Name of source for the first omics dataframe.\n",
      " |      df2_source (str): Name of source for the second omics dataframe.\n",
      " |      genes1 (str, or list or array-like of str, optional): Gene(s) for column(s) to select from df1_name. str if one key, list or array-like of str if multiple.\n",
      " |          Default of None will select entire dataframe.\n",
      " |      genes2 (str, or list or array-like of str, optional): Gene(s) for Column(s) to select from df2_name. str if one key, list or array-like of str if multiple.\n",
      " |          Default of None will select entire dataframe.\n",
      " |      how (str, optional): How to perform the join, acceptable values are from ['outer', 'inner', 'left', 'right']. Defaults to 'outer'.\n",
      " |      quiet (bool, optional): Whether to warn when inserting NaNs. Defaults to False.\n",
      " |      tissue_type (str): Acceptable values in [\"tumor\",\"normal\",\"both\"]. Specifies the desired tissue type for the dataframe. Defaults to \"both\".\n",
      " |      \n",
      " |      Returns:\n",
      " |      pandas.DataFrame: The selected columns from the two omics dataframes, joined into one dataframe.\n",
      " |  \n",
      " |  list_data_sources(self, source_filter='all')\n",
      " |      Prints which sources provide each data type and returns a DataFrame with this information.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      - source_filter (optional): Filter to select which sources are shown in the table.\n",
      " |          - \"all\" (default) returns all sources and data types.\n",
      " |          - If a list of sources is specified, only data types with data from these sources will be shown.\n",
      " |          - If a single source is specified, only data types from this source will be shown.\n",
      " |          This parameter is case-sensitive.\n",
      " |      \n",
      " |      Returns:\n",
      " |      - pandas.DataFrame: A DataFrame where each row represents a data type. The columns are 'Data type' and 'Available sources'.\n",
      " |      \n",
      " |      Raises:\n",
      " |      - InvalidParameterError: If a source specified in the source_filter is not a valid source for the dataset.\n",
      " |  \n",
      " |  list_definitions(self)\n",
      " |      Print all terms defined in the dataset's list of definitions.\n",
      " |  \n",
      " |  multi_join(self, join_dict: dict, mutations_filter: list = None, flatten: bool = False, levels_to_drop: list = [], how: str = 'outer', tissue_type: str = 'both') -> pandas.core.frame.DataFrame\n",
      " |      Joins multiple dataframes into a single dataframe based on the join_dict.\n",
      " |      \n",
      " |      Parameters:\n",
      " |      - join_dict (dict): A dictionary defining the dataframes and columns to join. Keys represent the names of dataframes. Values are lists of column names corresponding to each dataframe.\n",
      " |      - mutations_filter (list, optional): List of mutations to prioritize when filtering out multiple mutations, in order of priority. If None, all mutations are included.\n",
      " |      - flatten (bool, optional): If True, multiindexes in the dataframe will be flattened. Defaults to False.\n",
      " |      - levels_to_drop (list, optional): List of level names to drop from the dataframe. Defaults to an empty list (no levels dropped).\n",
      " |      - how (str, optional): Method of join. Accepts 'outer', 'inner', 'left', 'right'. Default is 'outer'.\n",
      " |      - tissue_type (str, optional): Type of tissue data to retrieve. Accepts 'tumor', 'normal', 'both'. Default is 'both'.\n",
      " |      \n",
      " |      Returns:\n",
      " |      - pandas.DataFrame: The resulting DataFrame after performing all joins specified in the join_dict.\n",
      " |      \n",
      " |      Example:\n",
      " |      ```python\n",
      " |      multi_join(\n",
      " |          join_dict={\n",
      " |              \"source1 dataframe1\": [\"column1\", \"column2\"],\n",
      " |              \"source2 dataframe2\": [\"column3\", \"column4\"]\n",
      " |          },\n",
      " |          mutations_filter=['nonsynonymous SNV'],\n",
      " |          flatten=True,\n",
      " |          levels_to_drop=[\"level1\", \"level2\"],\n",
      " |          how=\"outer\",\n",
      " |          tissue_type=\"tumor\"\n",
      " |      )\n",
      " |      ```\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from cptac.cancers.cancer.Cancer:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data and other attributes inherited from cptac.cancers.cancer.Cancer:\n",
      " |  \n",
      " |  NORMAL_ENDINGS = ('.N', '.C')\n",
      "\n"
     ]
    }
   ],
   "source": [
    "br = cptac.Brca()\n",
    "help(br)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3555ddf",
   "metadata": {},
   "source": [
    "Using help(br) we find the following function:\n",
    "\n",
    "get_proteomics(self, source: str = None, tissue_type: str = 'both', imputed: bool = False) -> pandas.core.frame.DataFrame\n",
    " |      Get the proteomics dataframe from the specified data source."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "cbe11d96",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Downloading BRCA_proteomics_gene_abundance_log2_reference_intensity_normalized_Tumor.txt.gz: 100%|██████████| 9.48M/9.48M [00:09<00:00, 1.02MB/s]  \n",
      "Downloading gencode.v34.basic.annotation-mapping.txt.gz: 100%|██████████| 1.75M/1.75M [00:02<00:00, 610kB/s] \n"
     ]
    }
   ],
   "source": [
    "br_prot = br.get_proteomics(\"bcm\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f737e16b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bcm': <cptac.cancers.bcm.bcmbrca.BcmBrca at 0x1186d0d90>,\n",
       " 'broad': <cptac.cancers.broad.broadbrca.BroadBrca at 0x10ebed6d0>,\n",
       " 'mssm': <cptac.cancers.mssm.mssm.Mssm at 0x15610ac90>,\n",
       " 'umich': <cptac.cancers.umich.umichbrca.UmichBrca at 0x157f2fd90>,\n",
       " 'washu': <cptac.cancers.washu.washubrca.WashuBrca at 0x158043690>,\n",
       " 'harmonized': <cptac.cancers.harmonized.harmonized.Harmonized at 0x158043d10>}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "br._sources"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26cde403",
   "metadata": {},
   "source": [
    "washu, mssm, and harmonized did not have proteomics data. bcm is the first one that does. However, it doesn't have somatic mutation data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5758cfc9",
   "metadata": {},
   "outputs": [
    {
     "ename": "DataTypeNotInSourceError",
     "evalue": "The umich source does not have somatic_mutation data for brca cancer.",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mDataTypeNotInSourceError\u001b[39m                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[21]\u001b[39m\u001b[32m, line 3\u001b[39m\n\u001b[32m      1\u001b[39m source = \u001b[33m\"\u001b[39m\u001b[33mumich\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      2\u001b[39m br_prot = br.get_proteomics(source)\n\u001b[32m----> \u001b[39m\u001b[32m3\u001b[39m mutations = \u001b[43mbr\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_somatic_mutation\u001b[49m\u001b[43m(\u001b[49m\u001b[43msource\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/anaconda3/envs/bio362/lib/python3.11/site-packages/cptac/cancers/cancer.py:172\u001b[39m, in \u001b[36mCancer.get_somatic_mutation\u001b[39m\u001b[34m(self, source, tissue_type, imputed)\u001b[39m\n\u001b[32m    170\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mget_somatic_mutation\u001b[39m(\u001b[38;5;28mself\u001b[39m, source: \u001b[38;5;28mstr\u001b[39m=\u001b[38;5;28;01mNone\u001b[39;00m, tissue_type: \u001b[38;5;28mstr\u001b[39m=\u001b[33m\"\u001b[39m\u001b[33mboth\u001b[39m\u001b[33m\"\u001b[39m, imputed: \u001b[38;5;28mbool\u001b[39m=\u001b[38;5;28;01mFalse\u001b[39;00m) -> pd.DataFrame:\n\u001b[32m    171\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"Get the somatic mutation dataframe from the specified data source.\"\"\"\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m172\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mget_dataframe\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43msomatic_mutation\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msource\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtissue_type\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimputed\u001b[49m\u001b[43m=\u001b[49m\u001b[43mimputed\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/anaconda3/envs/bio362/lib/python3.11/site-packages/cptac/cancers/cancer.py:694\u001b[39m, in \u001b[36mCancer.get_dataframe\u001b[39m\u001b[34m(self, data_type, source, tissue_type, imputed)\u001b[39m\n\u001b[32m    691\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m source \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m._sources:\n\u001b[32m    692\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m DataSourceNotFoundError(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mData source \u001b[39m\u001b[38;5;132;01m{\u001b[39;00msource\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m not found for the \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m._cancer_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m dataset.\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m694\u001b[39m df = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_sources\u001b[49m\u001b[43m[\u001b[49m\u001b[43msource\u001b[49m\u001b[43m]\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_df\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata_type\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    696\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m tissue_type == \u001b[33m\"\u001b[39m\u001b[33mnormal\u001b[39m\u001b[33m\"\u001b[39m:\n\u001b[32m    697\u001b[39m     df = \u001b[38;5;28mself\u001b[39m._normal_only(df)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/anaconda3/envs/bio362/lib/python3.11/site-packages/cptac/cancers/source.py:77\u001b[39m, in \u001b[36mSource.get_df\u001b[39m\u001b[34m(self, df_type)\u001b[39m\n\u001b[32m     74\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m df_type \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m._data:\n\u001b[32m     75\u001b[39m     \u001b[38;5;66;03m# check to see if the df type requested is availabe from this source\u001b[39;00m\n\u001b[32m     76\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m df_type \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m.load_functions:\n\u001b[32m---> \u001b[39m\u001b[32m77\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m DataTypeNotInSourceError(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mThe \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m.source\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m source does not have \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdf_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m data for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m.cancer_type\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m cancer.\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m     78\u001b[39m     \u001b[38;5;28mself\u001b[39m.load_functions[df_type]()\n\u001b[32m     79\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._data[df_type]\n",
      "\u001b[31mDataTypeNotInSourceError\u001b[39m: The umich source does not have somatic_mutation data for brca cancer."
     ]
    }
   ],
   "source": [
    "source = \"umich\"\n",
    "br_prot = br.get_proteomics(source)\n",
    "mutations = br.get_somatic_mutation(source)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35210199",
   "metadata": {},
   "source": [
    "Actually, it looks like none of these sources have both proteomic data and somatic mutation data. This is fatal."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio362",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
